src/pagination.rs

```diff
@@ -0,0 +1,275 @@
+        }
+        // list of available indexes, in `tags` like cases
+        if let Some(indexes) = self.indexes {
+            object.insert(
+                "indexes".into(),
+                liquid::value::Value::Array(
+                    indexes.iter().map(liquid::value::Value::scalar).collect(),
+                ),
+            );
+        }
+        object.insert(
+            "index".into(),
+            liquid::value::Value::scalar(self.index as i32),
+        );
+        object.insert(
+            "index_permalink".into(),
+            liquid::value::Value::scalar(self.index_permalink),
+        );
+        if let Some(previous_index_permalink) = self.previous_index_permalink {
+            object.insert(
+                "previous_index".into(),
+                liquid::value::Value::scalar(self.previous_index as i32),
+            );
+            object.insert(
+                "previous_index_permalink".into(),
+                liquid::value::Value::scalar(previous_index_permalink),
+            );
+        }
+        if let Some(next_index_permalink) = self.next_index_permalink {
+            object.insert(
+                "next_index".into(),
+                liquid::value::Value::scalar(self.next_index as i32),
+            );
+            object.insert(
+                "next_index_permalink".into(),
+                liquid::value::Value::scalar(next_index_permalink),
+            );
+        }
+        object.insert(
+            "first_index_permalink".into(),
+            liquid::value::Value::scalar(self.first_index_permalink),
+        );
+        object.insert(
+            "last_index_permalink".into(),
+            liquid::value::Value::scalar(self.last_index_permalink),
+        );
+        object.insert(
+            "total_indexes".into(),
+            liquid::value::Value::scalar(self.total_indexes as i32),
+        );
+        object.insert(
+            "total_pages".into(),
+            liquid::value::Value::scalar(self.total_pages as i32),
+        );
+        object
+    }
+}
+
+pub fn generate_paginators(
+    doc: &mut Document,
+    posts_data: &[liquid::value::Value],
+) -> Result<Vec<Paginator>> {
+    let config = doc
+        .front
+        .pagination
+        .as_ref()
+        .expect("Front should have pagination here.");
+    let mut all_posts: Vec<_> = posts_data.iter().collect();
+    match config.include {
+        Include::All => {
+            sort_posts(&mut all_posts, &config);
+            create_all_paginators(&all_posts, &doc, &config)
+        }
+        Include::None => {
+            unreachable!("PaginationConfigBuilder should have lead to a None for pagination.")
+        }
+    }
+}
+
+fn extract_value<'a>(a: &'a liquid::value::Value, key: &str) -> Option<&'a liquid::value::Scalar> {
+    let key = liquid::value::Value::scalar(key.to_owned());
+    let key = key.as_scalar().expect("Should have scalar");
```

Comment:
Why not create a Scalar directly?

Reply:
When trying this:
```rust
   let key = liquid::value::Scalar::new(key);
    a.get(key).and_then(|sort_key| sort_key.as_scalar())
```

I got:
```rust
error[E0277]: the trait bound `&liquid::value::Scalar: std::convert::From<liquid::value::Scalar>` is not satisfied
   --> src\pagination.rs:181:7
    |
181 |     a.get(key).and_then(|sort_key| sort_key.as_scalar())
    |       ^^^ the trait `std::convert::From<liquid::value::Scalar>` is not implemented for `&liquid::value::Scalar`
    |
    = help: the following implementations were found:
              <liquid::value::Scalar as std::convert::From<f64>>
              <liquid::value::Scalar as std::convert::From<std::string::String>>
              <liquid::value::Scalar as std::convert::From<&'a std::string::String>>
              <liquid::value::Scalar as std::convert::From<&'static str>>
            and 3 others
    = note: required because of the requirements on the impl of `std::convert::Into<&liquid::value::Scalar>` for `liquid::value::Scalar`
```